src/Raven.Server/Documents/PeriodicBackup/EncryptedBackup.cs

```diff
@@ -0,0 +1,298 @@
+        private readonly byte[] _plainTextBufer = new byte[DefaultBufferSize];
+        private Memory<byte> _plainTextWindow = Memory<byte>.Empty;
+
+        public unsafe DecryptingXChaCha20Oly1305Stream(Stream inner, byte[] key)
+        {
+            if (key.Length != (int)Sodium.crypto_secretstream_xchacha20poly1305_keybytes())
+                throw new InvalidOperationException($"The size of the key must be " +
+                                                    $"{(int)Sodium.crypto_secretstream_xchacha20poly1305_keybytes()} bytes, " +
+                                                    $"but was {key.Length} bytes.");
+
+            _inner = inner;
+            _pullState = new byte[(int)Sodium.crypto_secretstream_xchacha20poly1305_statebytes()];
+            var headerbytes = (int)Sodium.crypto_secretstream_xchacha20poly1305_headerbytes();
+            var header = stackalloc byte[headerbytes];
+            
+            var readingHeader = new Span<byte>(header, headerbytes);
+            while (readingHeader.IsEmpty == false)
+            {
+                var read = _inner.Read(readingHeader);  
+                if (read == 0)
+                    throw new EndOfStreamException("Wrong or corrupted file");
+
+                readingHeader = readingHeader.Slice(read);
+            }
+            fixed (byte* ps = _pullState, pKey = key)
+            {
+                if (Sodium.crypto_secretstream_xchacha20poly1305_init_pull(ps, header, pKey) != 0)
+                    throw new CryptographicException("Wrong key or corrupted file");
+            }
+        }
+
+        public override void Flush()
+        {
+            throw new NotSupportedException();
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (TryReadFromBuffer(buffer, offset, count, out int read))
+                return read;
+
+            while (read < _encryptedBuffer.Length)
+            {
+                int rc = _inner.Read(_encryptedBuffer, read, _encryptedBuffer.Length - read);
+                if (rc == 0)
+                    break;
+                read += rc;
+            }
+
+            if (read == 0)
+                return 0;
+
+            DecryptBuffer(read);
+
+            return Read(buffer, offset, count);
+        }
+
+        private unsafe void DecryptBuffer(int read)
+        {
+            fixed (byte* ciphertextPtr = _encryptedBuffer, statePtr = _pullState, plain = _plainTextBufer)
+            {
+                ulong s;
+                if (Sodium.crypto_secretstream_xchacha20poly1305_pull(statePtr, plain, &s, null, ciphertextPtr, (ulong)read, null, 0) != 0)
+                    throw new CryptographicException("Wrong or corrupted file");
```

Comment:
Explain failure to decrypt

Reply:
In particular, mention that the key might be wrong.
Probably need this on other errors as well.
Of course, don't include the key in the error

Reply:
Test this with the wrong file, we must provide good errors here for users